### PR TITLE
Add blog post on OTel context propagation

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes FerretDB/engineering#185.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+}
+```
+
+The strings `traceID` and `spanID` are hex-encoded strings that represent the corresponding trace and span identifiers.
+For example, the JSON above represents the following data:
+
+```go
+traceID := [16]byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef, 0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+spanID := [8]byte{0xfe, 0xdc, 0xba, 0x98, 0x76, 0x54, 0x32, 0x10}
+```
```
Comment (AlekSi) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:56]: I think that representation makes thing more confusing coming after JSON above, not more clear

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    comment, _ := json.Marshal(map[string]interface{}{
+        "ferretDB": traceContext,
+    })
+
+    var result bson.D
+    _ = collection.FindOne(findCtx, filter, options.FindOne().SetComment(string(comment))).Decode(&result)
+    findSpan.End()
+```
```
Comment (AlekSi) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:191]: That code also should be reduced to relevant snippets and made more tights. Right now, code dominates the message of the blog post; it should not.

**Review by AlekSi — COMMENTED** [review]
I think we should mention our other findings such as https://www.w3.org/TR/trace-context-protocols-registry/

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        sdktrace.WithBatcher(exporter),
+        sdktrace.WithResource(sdkresource.NewSchemaless(
+            semconv.ServiceName("test-app"),
+        )),
+    )
+
+    otel.SetTracerProvider(tp)
+
+    defer func() {
+        shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:115]: To let people run it without the need to redo things.

[the remark this answers is not in the document]
Comment (AlekSi) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:115]: But why let them do it? What's the benefit? I think most people will be interested in the general topic and direction, a few people will be adopting snippets, but who would use the full example and why?

To put it another way, this is not a blog post about programming in Go and best practices such as graceful shutdown with context.

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
+---
+slug: otel-context-propagation-in-ferretdb
+title: 'OpenTelemetry context propagation in FerretDB'
+authors: [elena]
+description: >
+  In this blog post, we demonstrate how to pass tracing context to queries in FerretDB using OpenTelemetry.
+tags: [observability]
+---
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:10]:
```suggestion
In distributed systems, achieving reliability depends on various factors, one of which is effective observability.
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+slug: otel-context-propagation-in-ferretdb
+title: 'OpenTelemetry context propagation in FerretDB'
+authors: [elena]
+description: >
+  In this blog post, we demonstrate how to pass tracing context to queries in FerretDB using OpenTelemetry.
+tags: [observability, opentelemetry]
+---
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
+OpenTelemetry (OTel) has emerged as a standard for distributed tracing, but passing context to databases remains a significant challenge,
```
[the remark this answers is not in the document]
Comment (Fashander) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:11]: maybe tracing context or data?

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 4 above)
Comment (AlekSi) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:10]: Let's fix content first, then review style and such

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+particularly with document databases like MongoDB.
+At FerretDB, we're committed to addressing this challenge.
+
+<!--truncate-->
+
+While adopting OpenTelemetry in application development is relatively straightforward across various programming languages,
+the process becomes more complex when passing context to databases.
+Most databases don't natively support tracing-related context.
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:19]: "tracing-related context" here may be more suitable to use throughout the post

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
+OpenTelemetry (OTel) has emerged as a standard for distributed tracing, but passing context to databases remains a significant challenge,
+particularly with document databases like MongoDB.
+At FerretDB, we're committed to addressing this challenge.
+
+<!--truncate-->
+
+While adopting OpenTelemetry in application development is relatively straightforward across various programming languages,
+the process becomes more complex when passing context to databases.
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md]: Removed

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 1 above)
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md]: removed

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 6 above)
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md]: rephrased the whole sentence

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+  In this blog post, we demonstrate how to pass tracing context to queries in FerretDB using OpenTelemetry.
+tags: [observability]
+---
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
+OpenTelemetry (OTel) has emerged as a standard for distributed tracing, but passing tracing data to databases remains a significant challenge,
+particularly with document databases like MongoDB.
+At FerretDB, we're committed to addressing this challenge.
+
+<!--truncate-->
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:15]: Let's have the "truncate" after the first sentence.

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+Context propagation, a concept that enables the tracking of requests as they move through different services, is explained by OpenTelemetry
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:17]: We should have at least 2 header sections in the post. If possible, keep the intro section within 120 words (2.5 paragraphs) before the next section.

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md]: fixed

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 2 above)
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:132]: fixed

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+Now, if we run the application, we will see that the spans are linked and the `FindCustomer` span is a child span created on the FerretDB side:
+
+[![Trace with propagation](/img/blog/ferretdb-otel/with-propagation.png)](/img/blog/ferretdb-otel/with-propagation.png)
+
+This approach gives more insights into the FerretDB's behavior and helps to understand the exact query lifecycle,
+making it easier to diagnose and understand performance issues or unexpected behavior.
+
+While this solution isn't perfect due to the limitations discussed, it is a step towards better context propagation in document databases.
+
+In conclusion, we believe that passing context to document databases is an important part of making them more observable.
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:143]:
```suggestion
## Conclusion

We believe that passing context to document databases is an important part of making them more observable.
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 8 above)
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:15]: Shouldn't we give the lead paragraph before `truncate`? I think, this is the lead paragraph :)

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
+---
+slug: otel-context-propagation-in-ferretdb
+title: 'OpenTelemetry context propagation in FerretDB'
+authors: [elena]
+description: >
+  In this blog post, we demonstrate how to pass tracing context to queries in FerretDB using OpenTelemetry.
+tags: [observability]
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:7]:
```suggestion
tags: [observability, product, open source]
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 9 above)
Comment (rumyantseva) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:19]: added some headers

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 8 above)
Comment (Fashander) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:15]: Not the lead paragraph, just like a sentence or 2 is okay.
Screenshot 2024-08-13 at 10 02 41
uce the

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+_ = collection.FindOne(findCtx, filter).Decode(&result)
+
+findSpan.End()
+```
+
+In this setup, we create two spans: one for inserting a document and another for finding it.
+
+If we run this application, we will see that the spans for the `InsertCustomer` and `FindCustomer` operations are created:
+
+[![Trace without propagation](/img/blog/ferretdb-otel/without-propagation.png)](/img/blog/ferretdb-otel/without-propagation.png)
```
Comment (chilagrow) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:112]: I think the format should be like `Trace without propagatio`.

Clicking image currently behaves weird. Also check other image

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+## Conclusion
+
+We believe that passing context to document databases is an important part of making them more observable.
+We hope that the community will come up with a standard way to pass context to databases, and we are looking forward to
+contributing to this effort.
+
+We'd love to hear your thoughts on this approach.
+Have you implemented similar context propagation strategies in your projects?
+Please feel free to reach out to us to share your experiences or ask questions!
```
Comment (chilagrow) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:160]: @Fashander should we have a link to how we can be reached?

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    "ferretDB": traceContext,
+})
+
+var result bson.D
+_ = collection.FindOne(findCtx, filter, options.FindOne().SetComment(string(comment))).Decode(&result)
+
+findSpan.End()
+```
+
+Now, if we run the application, we will see that the spans are linked and the `FindCustomer` span is a child span created on the FerretDB side:
```
Comment (chilagrow) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:143]: Possibly I'm misunderstanding, isn't FindCustomer now a parent span? FYI, I see usage of wording _parent_ instead of _child_ in docs such as https://opentelemetry.io/docs/concepts/context-propagation/ so we may want do the same.

**Review by chilagrow — COMMENTED** [review]
Awesome read

**Review by chilagrow — COMMENTED** [review]
See comments :hugs:

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 12 above)
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:112]:
```suggestion
![Trace without propagation](/img/blog/ferretdb-otel/without-propagation.png)
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+[![Trace with propagation](/img/blog/ferretdb-otel/with-propagation.png)](/img/blog/ferretdb-otel/with-propagation.png)
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:145]:
```suggestion
![Trace with propagation](/img/blog/ferretdb-otel/with-propagation.png)
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 13 above)
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:160]:
```suggestion
[Please feel free to reach out to us to share your experiences or ask questions](https://docs.ferretdb.io/#community)!
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+Let's consider a SQLCommenter-like approach for MongoDB.
+We can leverage MongoDB's existing features to attach metadata to queries.
+One such feature is the comment field, which allows additional information to be included in MongoDB queries.
+This field doesn't influence query logic but can be used to provide valuable context.
+
+At FerretDB, we want to empower users to pass such context.
+We decided to utilize the `comment` field to do it.
+We parse the content of the `comment` field, and if it's a JSON document with a `ferretDB` key, we check for tracing data.
+If the tracing data is present, FerretDB sets the parent's context when a span for the current operation is created.
```
Comment (AlekSi) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:49]: I guess it should be "parent context", not "parent's context". And probably we could rephrase it. For example, "FerretDB sets the parent context for the span created for the current operation" or something

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 14 above)
Comment (AlekSi) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:143]: I think it should be "child", but it would help to clarify child _of what_

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+making it easier to diagnose and understand performance issues or unexpected behavior.
+
+While this solution isn't perfect due to the limitations discussed, it is a step towards better context propagation in document databases.
+
+## Conclusion
+
+We believe that passing context to document databases is an important part of making them more observable.
+We hope that the community will come up with a standard way to pass context to databases, and we are looking forward to
+contributing to this effort.
+
```
Comment (AlekSi) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:157]: https://github.com/FerretDB/FerretDB/pull/4512#pullrequestreview-2235093373 might be somewhere there

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+Not all operations support the comment field.
+For instance, operations like `insert` or `listCollections` do not support it.
+
+Another challenge with the `comment` field is its handling in some MongoDB drivers, where it is typed as a string.
+This means that, in principle, any string can be passed.
+In our scenario, where we allow FerretDB to receive and parse
+tracing data through the `comment` field, the JSON must be correctly encoded by the driver and then decoded by FerretDB.
+
+A more robust solution would involve having a dedicated field for passing request-related context,
+and it would be ideal to establish a standard for such fields.
```
Comment (AlekSi) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:76]: https://github.com/FerretDB/FerretDB/pull/4512#pullrequestreview-2235093373 might be somewhere there

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/4512#pullrequestreview-2235093373 was ignored

**Review by AlekSi — COMMENTED** [review]
See above

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 14 above)
Comment (chilagrow) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:143]: Referencing with a name would help :+1:

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+For instance, it could be a BSON document with particular tracing-related fields.
+This would provide a more reliable method for passing context to the database.
+
+## Example application
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:80]:
```suggestion
### Example application
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+## Conclusion
+
+We believe that passing context to document databases is an important part of making them more observable.
+We hope that the community will come up with a standard way to pass context to databases, and we are looking forward to
+contributing to this effort.
+
+We'd love to hear your thoughts on this approach.
+Have you implemented similar context propagation strategies in your projects?
+[Please feel free to [reach out to us here](https://docs.ferretdb.io/#community) to share your experiences or ask questions!
```
Comment (Fashander) [inline website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:160]:
```suggestion
Please feel free to [reach out to us here](https://docs.ferretdb.io/#community) to share your experiences or ask questions!
```

File: website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md (hunk 8 above)
Comment (Fashander) [reply website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md:17]: The "truncate" should be after the first sentence so it's not too wordy on the homepage. Don't worry about the intro; the metadata description is picked up for search engines. This one is just for the FerretDB blog.

**Review by Fashander — COMMENTED** [review]
It's a good read! I left a few suggestions, and it should be food to go from there

File: website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+```json
+{
+  "ferretDB": {
+    "traceID": "1234567890abcdef1234567890abcdef",
+    "spanID": "fedcba9876543210"
+  }
+}
+```
+
+The strings `traceID` and `spanID` are hex-encoded strings that represent the corresponding trace and span identifiers.
```
Comment (chilagrow) [inline website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:62]: I think `hex-encoded strings` is no longer the case from above examples. Unless its about how it's implemented in FerretDB :thinking:

File: website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+traceContext := struct {
+    TraceID string `json:"traceID"`
+    SpanID  string `json:"spanID"`
+}{
+    TraceID: traceID,
+    SpanID:  spanID,
+}
+
+comment, _ := json.Marshal(map[string]interface{}{
+    "ferretDB": traceContext,
```
Comment (chilagrow) [inline website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:134]: I don't know below would work it's simpler to read.

```suggestion
comment, _ := json.Marshal(map[string]interface{}{
    "ferretDB": map[string]interface{}{
        "traceID": traceID,
        "spanID": spanID,
    },
```

**Review by chilagrow — APPROVED** [review]
Just tiny comments from me. They are optional as I don't think it's so important.

**Review by noisersup — DISMISSED** [review]
LGTM

**Review by chilagrow — DISMISSED** [review]
great work!

**Review by noisersup — DISMISSED** [review]

File: website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md
Comment (chilagrow) [inline website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:1]: Let's change the publication date :hugs:

Comment (rumyantseva) [reply website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:1]: @AlekSi do you know the date yet?

Comment (AlekSi) [reply website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:1]: As discussed on DS, we can't release a blog post that describes a functionality as done that is not, in fact, released yet, and we don't know a release date

Comment (rumyantseva) [reply website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md:1]: Ok, then if the release happens when I'm on vacation, please feel free to change the date and commit any other necessary modifications.
